### DISCUSSION TITLE
[7 Days To Die Modding] - Change executable to run sdtdserver with Undead Legacy

### DISCUSSION THREAD

**Discussion by vinanrra** [GameServerManagers/LinuxGSM#3755] [Q&A] [answered]
7 Days To Die - sdtdserver

## Can you link to any tutorials or guides?

I'm trying to modify the linuxgsm start command to works with [Undead Legacy mod](https://gitlab.com/Subquake/UndeadLegacyExperimental).
This mod install various things and need to run his own script to start the server `run_bepinex_server.sh`
I found this [line](https://github.com/GameServerManagers/LinuxGSM/blob/master/lgsm/config-default/config-lgsm/sdtdserver/_default.cfg#L162) in _default.cfg for sdtdserver

`executable="./7DaysToDieServer.x86_64"`

If i swap that line to:

`executable="./run_bepinex_server.sh"`

and remove the startparameters (because the script have already all start parameters:

`startparameters=""`

Will run the custom script?

**Comment by vinanrra** [comment]
Thanks for moving the issue, didn't know that Discussions exist

PD: I will test it later, but wanted to know if that works, because i can't test it right now

**Comment by vinanrra** [comment, marked as answer]
Its working here are my edits to sdtdserver.cfg
```cfg

##################################
####### Instance Settings ########
##################################
# PLACE INSTANCE SETTINGS HERE
## These settings will apply to a specific instance.
startparameters=""
executable="./run_bepinex_server.sh"
```

**Reply by XBadPlayerX** [reply]
Can you go over everything you did to get Undead Legacy going?

I am having these same issues and I have done what you posted above and it still isn't working.

Thanks in advance.

**Reply by Jellypowered** [reply]
Edit the contents of /lgsm/config-lgsm/sdtdserver/sdtdserver.cfg and replace with the above. Should work.

**Reply by XBadPlayerX** [reply]
> ```haproxy-config
> startparameters=""
> executable="./run_bepinex_server.sh"
> ```

Thanks for pointing that out!

Unfortunately I am still getting the fail error.
```
XXXXX@XXXXXX:~$ ./sdtdserver start
[ FAIL ] Starting sdtdserver: Unable to start BadPlayer Gaming Test Server
```

**Reply by Jellypowered** [reply]
Can you run the server directly from the serverfiles dir using ./run_bepinex_server.sh ?

**Reply by XBadPlayerX** [reply]
No, it says `Please open run.sh in a text editor and configure executable name.`

**Reply by XBadPlayerX** [reply]
Oh, wait.

Yes, it ran, but I don't believe it started.

**Reply by XBadPlayerX** [reply]
Here is what was spit out.

https://pastebin.com/hu8mWvfm

**Reply by Jellypowered** [reply]
Looks like you may have a instance already running, The log says none of the ports are available. 
You can check with htop (sudo apt install htop) 

Look for the server running already. 
![image](https://user-images.githubusercontent.com/30215109/187053581-3a40a17d-ea3b-4bba-8e48-1e5f58c2f0ed.png)

**Reply by XBadPlayerX** [reply]
So, I did have another server running that shouldn't have been.

I swear they auto start sometimes.

But, the test server is not running.

A little back story.

I have 4 servers running, all on separate users in Ubuntu.

I am using the common.cfg file, so is that where I would edit which executable to use?

**Reply by Jellypowered** [reply]
If you're running multiple 7dtd servers, make sure you have a 3 port space available. 
Whatever port you specify the server will want access to 3 ports in total. 
Vanilla: 26900-26902
UL: 26903-26905

Also, I saw a telnet error, Make sure you increment the telnet and web admin ports for each server as well. These only need 1 port. 
Vanilla serverfiles/sdtdserver.xml: 
 ```
<property name="ControlPanelPort"                               value="8080"/>
  <property name="TelnetPort"                                             value="8081"/>
```
UL serverfiles/sdtdserver.xml: 
```
 <property name="ControlPanelPort"                               value="8082"/>
  <property name="TelnetPort"                                             value="8083"/>
```

Since you're hosting from the same machine, port conflicts will halt things (in theory anyway 😅). 

Also, I noticed from your log file, You're using Navezgane instead of RWG: 
UL will not load any of it's custom POI's in a premade map.  Make a RWG! 

The only thing I had to edit was /lgsm/config-lgsm/sdtdserver/sdtdserver.cfg as shown when we started this discussion. I recommend purging and reinstalling. Something isn't jiving.  Get the vanilla server going, to ensure the install was good. Then install UL, and edit /lgsm/config-lgsm/sdtdserver/sdtdserver.cfg 

Good luck!!! You got this!

**Reply by XBadPlayerX** [reply]
Thanks.

I know about the ports. These servers have been running for a couple years now.

I am just trying to get Undead Legacy working.

I think it is something to do with Bepinex, but I don't know enough about it.

I will try again tomorrow.

Thanks for your help.

**Reply by Jellypowered** [reply]
Also! I don't know if this will make a difference: But UL is only currently compatible with A20.5 right now. 
Might need to use steamcmd to change betas as 20.6 is live.

**Reply by XBadPlayerX** [reply]
Yeah, client and server are set to 20.5.

**Reply by XBadPlayerX** [reply]
I uninstalled the server, validated the files with 20.5.
Reinstalled Undead Legacy, changed the last line in the `MonoBleedingEdge` config and changed the executable line in `config-lgsm/sdtdserver/common.cfg` file, as that's were my other config data is.

Here are the script logs
```
Aug 28 00:39:40.145 sdtdserver: START: WARN: Permissions issue found
Aug 28 00:39:40.650 sdtdserver: START: INFO: The following file is not executable:
Aug 28 00:39:40.651 sdtdserver: START: INFO: /home/test/serverfiles/run_bepinex_server.sh
Aug 28 00:39:41.154 sdtdserver: START: INFO: Applying chmod u+x,g+x 664
Aug 28 00:39:43.819 sdtdserver: START: INFO: Using anonymous Steam login
Aug 28 00:39:44.744 sdtdserver: START: INFO: Rotating log files
Aug 28 00:39:45.803 sdtdserver: START: FATAL: Unable to start BadPlayer Gaming Test Server
Aug 28 00:39:45.810 sdtdserver: START: INFO: LinuxGSM version: v22.1.0
Aug 28 00:39:45.812 sdtdserver: START: FATAL: core_exit.sh exiting with code: 1
Aug 28 00:40:07.506 sdtdserver: STOP: ERROR: BadPlayer Gaming Test Server is already stopped
Aug 28 00:40:08.732 sdtdserver: START: INFO: Using anonymous Steam login
Aug 28 00:40:09.603 sdtdserver: START: INFO: Rotating log files
```

And here are the console logs from the same times.
```
Cannot identify executable type (got directory)!
Please create an issue at https://github.com/BepInEx/BepInEx/issues.
/home/test/serverfiles/
Cannot identify executable type (got directory)!
Please create an issue at https://github.com/BepInEx/BepInEx/issues.
```

**Reply by Arcitec** [reply]
@XBadPlayerX The bug is caused by `run_bepinex_server.sh`

You need to DELETE this whole section from the top of that file:

```sh
if [ -n "$1" ]; then
    case $os_type in
        Linux*)
            executable_path="$1"
            ;;
        Darwin*)
            # Special case: allow to specify path to the executable within .app
            full_path_part=$(echo "$1" | grep "\.app/Contents/MacOS")
            if [ -z "$full_path_part" ]; then
                executable_name=$(basename "$1" .app)
                real_executable_name=$(defaults read "$1/Contents/Info" CFBundleExecutable)
                executable_path="$1/Contents/MacOS/${real_executable_name}"
            else
                executable_path="$1"
            fi
            ;;
    esac
fi
```

The issue is that LinuxGSM sends parameter to `run_bepinex_server.sh`, and that script in turn thinks that any parameter means that you told it the full server executable as the 1st parameter.

---

You should also do a few other changes:

Let's change their godawful "base directory" calculation, by doing the following change near the top of the file.

```sh
# Replace this:
a="/$0"; a=${a%/*}; a=${a#/}; a=${a:-.}; BASEDIR=$(cd "$a"; pwd -P)

# With this:
BASEDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
```

---

Lastly, change the final line as follows, to make it pass through all command line arguments it's been given by LinuxGSM (we also keep their hardcoded flags to disable EAC and GameSpark):

```sh
# Replace this:
"${executable_path}" -configfile=$config_file -noeac -nogs

# With this:
"${executable_path}" -noeac -nogs "$@"
```

Save the changes to the bepinex start script.

---

Now just do one more thing: Remove your custom LinuxGSM `startparameters=""` stuff from your lgsm config. Remove that. Make LinuxGSM send the executable parameters again. They will then be sent forwards into the bepinex script, which in turn sends it forwards into the game thanks to the changes I made.

**Reply by vinanrra** [reply]
> I uninstalled the server, validated the files with 20.5. Reinstalled Undead Legacy, changed the last line in the `MonoBleedingEdge` config and changed the executable line in `config-lgsm/sdtdserver/common.cfg` file, as that's were my other config data is.
> 
> Here are the script logs
> 
> ```
> Aug 28 00:39:40.145 sdtdserver: START: WARN: Permissions issue found
> Aug 28 00:39:40.650 sdtdserver: START: INFO: The following file is not executable:
> Aug 28 00:39:40.651 sdtdserver: START: INFO: /home/test/serverfiles/run_bepinex_server.sh
> Aug 28 00:39:41.154 sdtdserver: START: INFO: Applying chmod u+x,g+x 664
> Aug 28 00:39:43.819 sdtdserver: START: INFO: Using anonymous Steam login
> Aug 28 00:39:44.744 sdtdserver: START: INFO: Rotating log files
> Aug 28 00:39:45.803 sdtdserver: START: FATAL: Unable to start BadPlayer Gaming Test Server
> Aug 28 00:39:45.810 sdtdserver: START: INFO: LinuxGSM version: v22.1.0
> Aug 28 00:39:45.812 sdtdserver: START: FATAL: core_exit.sh exiting with code: 1
> Aug 28 00:40:07.506 sdtdserver: STOP: ERROR: BadPlayer Gaming Test Server is already stopped
> Aug 28 00:40:08.732 sdtdserver: START: INFO: Using anonymous Steam login
> Aug 28 00:40:09.603 sdtdserver: START: INFO: Rotating log files
> ```
> 
> And here are the console logs from the same times.
> 
> ```
> Cannot identify executable type (got directory)!
> Please create an issue at https://github.com/BepInEx/BepInEx/issues.
> /home/test/serverfiles/
> Cannot identify executable type (got directory)!
> Please create an issue at https://github.com/BepInEx/BepInEx/issues.
> ```

Forgot to reply, if you check my 7 days to die docker image at Mods/ you will see how the scripts handles all the edits for you
